### ISSUE TITLE
High Training Loss on very Small DBs with PBCs

### ISSUE THREAD

**Issue by JordD04** [atomistic-machine-learning/schnetpack#288]
I've been having trouble getting SchNetPack to perform as well as I would expect on my custom database (MAE Energy 1.5 eV / atom and MAE Force 0.5 eV / Å) so I decided to do some testing using databases that only have 10 structures (I just pick 10 structures at random from a full database and save them into a new database file). My understanding is that when training a NN on a very small dataset, we should expect terrible validation loss but a near-zero training loss.

I've been using spk_run.py to minimise user-error on my part and my args.json looks like this:
```
{
    "aggregation_mode": "sum",
    "batch_size": 100,
    "checkpoint_interval": 1,
    "contributions": null,
    "cuda": false,
    "cutoff": 10.0,
    "cutoff_function": "cosine",
    "datapath": "model/QM9_10_set2.db",
    "dataset": "custom",
    "derivative": null,
    "environment_provider": "simple",
    "features": 128,
    "force": null,
    "interactions": 6,
    "keep_n_checkpoints": 3,
    "log_every_n_epochs": 1,
    "logger": "csv",
    "lr": 0.0001,
    "lr_decay": 0.8,
    "lr_min": 1e-06,
    "lr_patience": 25,
    "max_epochs": 5000,
    "max_steps": null,
    "mode": "train",
    "model": "schnet",
    "modelpath": "model/",
    "n_epochs": 1000,
    "negative_dr": false,
    "num_gaussians": 50,
    "output_module": "atomwise",
    "overwrite": false,
    "parallel": false,
    "property": "energy_U0",
    "rho": {},
    "seed": null,
    "split": [
        7,
        3
    ],
    "split_path": null,
    "stress": null
}
```
For QM9, I had trouble getting a decent train loss with the default parameters from the benchmark (you'll notice I've specified a custom dataset in the args.json but I think I'm using the parameters that come with the QM9 benchmark) but I was able to consistently achieve a train loss of 0 when I increased the lr from 1e-4 to 1e-3.
The figure shows the train loss converging to zero.
![QM9-best](https://user-images.githubusercontent.com/1126249/118511209-9669f280-b729-11eb-8bac-1187bc714295.png)

I've tried training on energy_per_atom and total_energy (I defined this later property myself) and I got a training loss range of 0 - 0.404. I've tested with both lr=1e-4 to Lr=1e-3 but it didn't change very much.
The figure shows the train loss converging to zero.
![MatProj-best](https://user-images.githubusercontent.com/1126249/118511228-9bc73d00-b729-11eb-8117-a553a8039129.png)

I experimented with a few parameters for my custom dataset. The parameters that gave me the best results are as follows: lr = 1e-3, lr_decay = 0.8, lr_min = 1e-6, lr_patience= 50 and converged to a train loss over the range of 0.143 - 4.778 (units are in eV). Simply changing my lr from 1e-4 to 1e-3 gave worse results. 

I've also tried training on Nongnuch Artrith's TiO2 benchmark (xsf files converted to a SchNetPack database) but got even worse results. Here, an lr=1e-3 converged to train loss of a range between 0.114 and 6895. 
The figure shows the train loss in the worst case scenario.
![TiO2-worst](https://user-images.githubusercontent.com/1126249/118512067-64a55b80-b72a-11eb-8f3d-950fed025551.png)
Training on the full TiO2 database gives similarly terrible results (MAE energy = 35 eV / atom).

I've experimented with some of the other non-lr-related arguments and I'm not sure they're the problem. I don't think i'm being conservative with the architecture parameters or cfconv parameters. I don't think I even have enough RAM for the current parameters when scaling up to a full dataset.

I've seen another issue reported on [github](https://github.com/atomistic-machine-learning/schnetpack/issues/226) concerning training on the MatProj benchmark. Does SchNetPack not work with PBCs right now or am I making a mistake somewhere?

**Comment by Stefaanhess** [comment]
Hi @JordD04 ,

What environment-provider are you using for the PBC models? In the args-file it looks like the simple environment-provider is used, but you will need to use the torch env-provider.
I tried to reproduce your experiments with a database of SiO2, that I have been working with lately and the train-loss converges to very low values while the validation loss does not (as expected). Maybe you could also check if the energy values match the forces in you dataset. E.g. if you are maybe training on energy/atom and forces. This could be an issue, since the derivatives do not match then.

Best, Stefaan

**Comment by JordD04** [comment]
Hey Stefaan, thanks for getting back to me.
I did some testing following your recommendations.
Switching to the Torch environment has allowed me to consistently get Train Loss = 0 (or very close to) for my small MatProj db but I continue to struggle with my custom dataset. 
Using the same parameters as in the args.json above (but with the Torch environment, I've gotten a loss as high as 3.4.
With lr = 1E-3, I get a loss as high as 2.9, and if I step the lr_patience up to 50, I get a loss as high as 4.7. In all cases though, I got at least 1 run that converged to train loss = 0.
Interestingly, I find that the MatProj dataset converges much more slowly than the custom one. The Custom set takes around 1000 seconds to converge and the MatProj set takes between 2000 and 3500 seconds. You can see it here:
![Train_comparison](https://user-images.githubusercontent.com/1126249/118828733-6fd8c280-b8b5-11eb-9b94-8e1597d8b005.png)
I was worried I might be setting up my database incorrectly so I exported the contents of my small MatProj db as .res files and then turned it back into a database file. I was still able to get the train loss down to 0.
So the problem almost certainly isn't related to PBCs, but there must be something going wrong.

My full dataset has energies and forces, but I've just been training only on energies for these tests.

**Comment by Stefaanhess** [comment]
Hey @JordD04 ,
if you are training on energy/atom, you will also need to set `aggregation_mode=avg`. Since you are building your own database, I would check if `pbc=True` is set in the atoms objects of your database. Otherwise this will also cause problems for the neighbour selection.
Let me know, if this helps to solve the issue.

**Comment by JordD04** [comment]
Hey Stefaan,
I definitely have pbc set to true but I haven't tried `aggregation_mode=avg` yet. Your readthedocs says 
> Since energy is a property that depends on the total number of atoms we select --aggregation_mode sum
When should I be using sum and when should I be using avg?

**Comment by Stefaanhess** [comment]
This depends on the definition of energy in your dataset. For molecules this is usually the total energy of the structure, which is interpreted by the model as the sum over the local energy contributions of every atom. For pbc-structures this definition is not working well since the structure is infinite. So you could either define the energy as the sum over all atoms inside the cell, or as energy per atoms. If the energy is per cell `aggregation_mode=sum` is needed, because the local energies are summed up over the cell. If the energy is per atom avg contributions of every atom are needed and you will need to set `aggregation_mode=avg`.

**Comment by JordD04** [comment]
I was training on the energy of the cell before but I've defined an energy_per_atom property, set `aggregation_mode=avg` and then trained 5 more NN. The largest loss was 0.006, so this looks quite promising. I'm going to try and train on my full dataset and see if these new parameters improve my MAE energy and loss. Thanks for your help!

**Comment by Stefaanhess** [comment]
Sure, no problem! Let me know if it worked out well.

**Comment by JordD04** [comment]
Sorry for the delayed response. My cluster's GPUs are offline so I have to everything slowly on GPUs. I ran a few tests; I got the best results from the args.json below:
`{
    "aggregation_mode": "avg",
    "batch_size": 30,
    "checkpoint_interval": 1,
    "contributions": null,
    "cuda": false,
    "cutoff": 5.0,
    "cutoff_function": "cosine",
    "datapath": "model/third_gen_few_atoms_non_zero_forces_epa.db",
    "dataset": "custom",
    "derivative": "forces",
    "environment_provider": "torch",
    "features": 128,
    "force": null,
    "interactions": 3,
    "keep_n_checkpoints": 3,
    "log_every_n_epochs": 1,
    "logger": "csv",
    "lr": 0.0001,
    "lr_decay": 0.8,
    "lr_min": 1e-06,
    "lr_patience": 25,
    "max_epochs": 5000,
    "max_steps": null,
    "mode": "train",
    "model": "schnet",
    "modelpath": "model/",
    "n_epochs": 1000,
    "negative_dr": true,
    "num_gaussians": 20,
    "output_module": "atomwise",
    "overwrite": false,
    "parallel": true,
    "property": "energy_per_atom",
    "rho": {},
    "seed": null,
    "split": [
        114545,
        11455
    ],
    "split_path": null,
    "stress": null
}
`

After about 3 days of compute time (but still running) I've approached an energy MAE of 0.152 eV / atom, a force MAE of 0.180 eV / Å and a train loss of 0.049. Assuming that the error reported by Schnetpack is actually per atom, this is a big improvement over training on total energies with aggregation_mode sum but I think it still falls short of being an accurate model.
I embed the associated graphs below:
![schnet_energy_and_force](https://user-images.githubusercontent.com/1126249/119333127-6c608500-bc81-11eb-81ec-f30674552ccd.png)
![schnet_loss](https://user-images.githubusercontent.com/1126249/119333138-6ff40c00-bc81-11eb-814d-2d96080ea7ab.png)
You may notice that i've picked some slightly cheaper parameters than my prior args.json, notably cutoff, interactions, and num_gaussians. I tried to train with those parameters and after 20 hours on a 20-core haswell, it still hadn't computed a single iteration. I checked my current parameters on my set of 10 structures and was still able to get an appropriately low train loss.

At this point I imagine the issue is just that my batch size, interactions, cutoff, and num_guassians is too low, so I will have to wait  until I can get access to a GPU again?

**Comment by MrRobot2211** [comment]
Hi, maybe it will be useful to point out that in the paper the feature dimension for materials project is 64 instead of 128. Maybe by switching to 64 yo will reduce the overfitting, I am trying it myself.
